src/engine/execute.js

```diff
@@ -107,16 +107,161 @@ const handleReport = function (
+ * @param {object} cached default set of cached values
  */
-const FieldKind = {
-    NONE: 'NONE',
-    VARIABLE: 'VARIABLE',
-    LIST: 'LIST',
-    BROADCAST_OPTION: 'BROADCAST_OPTION',
-    DYNAMIC: 'DYNAMIC'
-};
+class BlockCached {
+    constructor (blockContainer, cached) {
+        /**
+         * Block operation code for this block.
+         * @type {string}
+         */
+        this.opcode = cached.opcode;
+
+        /**
+         * Original block object containing argument values for static fields.
+         * @type {object}
+         */
+        this.fields = cached.fields;
+
+        /**
+         * Original block object containing argument values for executable inputs.
+         * @type {object}
+         */
+        this.inputs = cached.inputs;
+
+        /**
+         * Procedure mutation.
+         * @type {?object}
+         */
+        this.mutation = cached.mutation;
+
+        /**
+         * Is the opcode a hat (event responder) block.
+         * @type {boolean}
+         */
+        this._isHat = false;
+
+        /**
+         * The block opcode's implementation function.
+         * @type {?function}
+         */
+        this._blockFunction = null;
+
+        /**
+         * Is the block function defined for this opcode?
+         * @type {boolean}
+         */
+        this._definedBlockFunction = false;
+
+        /**
+         * Is this block a block with no function but a static value to return.
+         * @type {boolean}
+         */
+        this._isShadowBlock = false;
+
+        /**
+         * The static value of this block if it is a shadow block.
+         * @type {?any}
+         */
+        this._shadowValue = null;
+
+        /**
+         * A copy of the block's fields that may be modified.
+         * @type {object}
+         */
+        this._fields = Object.assign({}, this.fields);
+
+        /**
+         * A copy of the block's inputs that may be modified.
+         * @type {object}
+         */
+        this._inputs = Object.assign({}, this.inputs);
+
+        /**
+         * An arguments object for block implementations. All executions of this
+         * specific block will use this objecct.
+         * @type {object}
+         */
+        this._argValues = {
+            mutation: this.mutation
+        };
+
+        const {runtime} = blockUtility.sequencer;
+
+        const {opcode, fields, inputs} = this;
+
+        // Assign opcode isHat and blockFunction data to avoid dynamic lookups.
+        this._isHat = runtime.getIsHat(opcode);
+        this._blockFunction = runtime.getOpcodeFunction(opcode);
+        this._definedBlockFunction = typeof this._blockFunction !== 'undefined';
+
+        const fieldKeys = Object.keys(fields);
+
+        // Store the current shadow value if there is a shadow value.
+        this._isShadowBlock = fieldKeys.length === 1 && Object.keys(inputs).length === 0;
+        this._shadowValue = this._isShadowBlock && fields[fieldKeys[0]].value;
```

Comment:
I think shadowBlock/shadowValue is not the best name for what this represents. It looks like this refers to any 'field' of a block, whereas shadows are specifically used for block fields that are 'pluggable'. Maybe rename this to fieldBlock and fieldValue?

Reply:
This is how shadow values are stored in blocks. They are are in blocks with no inputs and 1 field. These two members are specifically used in execute by a block of code that determines the block is a block containing a shadow value and reports the shadow value in the only field.

Reply:
Here's the comment from the relevant block in execute.

```js
// Hats and single-field shadows are implemented slightly differently
// from regular blocks.
// For hats: if they have an associated block function, it's treated as a
// predicate; if not, execution will proceed as a no-op. For single-field
// shadows: If the block has a single field, and no inputs, immediately
// return the value of the field.
```

Reply:
Instead of changing the names I made the `_isShadowBlock` containing an additional detail for whether its a shadow block or not. Now `_isShadowBlock` is true when the block can only refer to a block and false when it is not a shadow block.